### PR TITLE
lsp: resolve the analysis source root via manifest walk-up with closure fallback

### LINKED ISSUE #243: lsp: non-entry file opened standalone resolves imports against its own directory — false missing-import diagnostics on valid projects
Follow-up from the comprehensive review of PR #239 (LSP server). Line numbers reference the PR branch at review time.

**Severity: major** (reproduced by execution during review; known v1 trade-off, now confirmed user-visible)

Import paths are source-root-relative for every file in a closure (one `src_root` per walk; `project.rs` `enqueue_imports` uses `module_file_path(self.src_root, ...)`), but `RootDatabase` analyzes each open file as its own entry with **its own directory** as source root (`ide/ide-db/src/database.rs:14` area).

Repro: project `main.inf` (`use lib::a;`), `lib/a.inf` (`use lib::b;`), `lib/b.inf`. From the entry all 3 files resolve with 0 import problems. Opening `lib/a.inf` standalone resolves `lib::b` against `/proj/lib` as root, probing `/proj/lib/lib/b.inf` → 1 missing-import problem. Since the LSP server pushes diagnostics per open document from that document's own-entry analysis, any user who goto-defs or browses into a subdirectory file of a multi-directory project immediately gets false "file not found for import" squiggles (plus missing symbols / type errors) on a file the compiler accepts — the exact compiler/IDE disagreement the shared FileLoader seam was built to prevent.

This is the AD-4 limitation from the original design (open non-entry file analyzed as its own entry); manifest-aware entry mapping was noted as the forward path.

**Suggested fix**: resolve a project root for opened files (walk up to a workspace/manifest root, or reuse the closure of an already-analyzed entry that contains the file) before falling back to file-as-own-entry; at minimum suppress missing-import diagnostics for documents already present in another open entry's closure.

### PR DESCRIPTION
**Pull request by 0xGeorgii**
Fixes #243. **Stacked on #261** (`242-bug-fix-stale-analysis-unreadable-import`) — merge that first; GitHub will retarget this PR to `main` when the base branch is deleted.

A non-entry file opened standalone was analyzed with its own directory as source root, so root-relative imports resolved against the wrong root — false missing-import squiggles (and missing symbols) on files the compiler accepts. Repro from the issue: `main.inf` / `lib/a.inf` / `lib/b.inf`, opening `lib/a.inf` probed `/proj/lib/lib/b.inf`.

## Design (three tiers, first match wins)
1. **Manifest walk-up**: nearest-ancestor `Inference.toml`; source root = `<manifest_dir>/src`, mirroring how `infs` compiles `src/main.inf` (verified against `ProjectContext::entry_relative`). Nearest manifest wins; a malformed nearest manifest falls through rather than climbing past it — matching infs. Manifests created/edited after open are not observed in v1 (no file watcher); documented at the derivation site.
2. **Closure fallback**: reuse the source root of an already-analyzed open entry whose import closure contains the file (deterministic donor: lexicographically smallest entry path; a file is never its own donor).
3. **Own directory** — today's behavior, unchanged.

## Changes
- New minimal `inference::manifest` module in `core/inference` (walk-up + just-enough `[package]` validation) so IDE and CLI agree on manifest semantics without `ide-db` depending on `apps/infs` — #256 will extract a proper project-model crate later.
- `load_project_resilient_with_root` (explicit root); `load_project_resilient` delegates with `entry.parent()`, keeping the compiler path byte-identical.
- `FileAnalysis` records the root it used; `RootDatabase::analysis` resolves and threads it. Invalidation confirmed correct under the new root.
- Dependency note: `toml` was promoted to `[workspace.dependencies]` (same `1.1.2` already used by `apps/infs`) and wired into `core/inference` — no genuinely new third-party crate.

## Testing
- 14 manifest-helper unit tests (same-dir/ancestor/none/nearest-wins, under-src/outside-src, malformed TOML, missing/partial `[package]`, unknown-key forward-compat, directory-named-manifest).
- 7 ide-db integration tests: the issue repro under a manifest (0 import problems, `lib/b` symbols visible), nested-manifest nearest-wins, outside-src fall-through, closure-donor reuse + negative (unrelated entry doesn't donate), own-dir tier unchanged, didChange-under-manifest-root invalidates and recovers. Positive tests verified to fail on pre-fix behavior.
- `cargo test -p inference` (100 lib + 19 doctests) and `-p inference-ide-db` (43 lib + 30 integration) green; clippy `-D warnings` clean incl. downstream `inference-ide`/`inference-lsp`; full default-workspace `cargo test` green (44 binaries, 0 failures).

<h3>Confidence Score: 5/5</h3>

Safe to merge; the change is well-scoped and thoroughly tested, including the concurrent-eviction edge case flagged in the previous review.

The three-tier root resolution and sticky cache are each verified by integration tests, the compiler path is byte-identical (load_project_resilient still delegates with entry.parent()), invalidation is unchanged, and the only new third-party dependency (toml) was already in the workspace at the same version.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| core/inference/src/manifest.rs | New module: manifest walk-up, package validation, and source-root derivation; 14 unit tests cover same-dir, ancestor, nearest-wins, under-src, outside-src, malformed TOML, missing/partial package, forward-compat keys, and directory-named manifest. |
| ide/ide-db/src/database.rs | Core of the fix: three-tier source-root resolution with a sticky cache (source_roots map); close_document drops the cached root; close_document/invalidate interaction correctly evicts both donor and dependent analyses on shared-dep change. |
| ide/ide-db/src/analysis.rs | FileAnalysis gains a src_root field and compute() now takes an explicit src_root; source_root() accessor exposes it to the database for closure-donor reuse. |
| core/inference/src/project.rs | load_project_resilient now delegates to the new load_project_resilient_with_root; original behavior (entry.parent() as root) is byte-identical on the compiler path. |
| ide/ide-db/tests/database.rs | 7 new integration tests: issue repro, nested-manifest nearest-wins, outside-src fall-through, closure-donor reuse + negative, own-dir tier unchanged, didChange invalidation+recovery, and the concurrent-eviction sticky-root survival case. |
| Cargo.toml | toml 1.1.2 promoted to workspace dependencies (same version already used by apps/infs); wired into core/inference/Cargo.toml. |

<h3>Flowchart</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["analysis(path) called"] --> B{analysis memoized?}
    B -- Yes --> R["return cached FileAnalysis"]
    B -- No --> C["resolve_source_root(path)"]

    C --> D{sticky root cached\nin source_roots?}
    D -- Yes --> USE["use cached root"]

    D -- No --> E["Tier 1: manifest_source_root(path)"]
    E --> F{Inference.toml found\n& file under src/?}
    F -- Yes --> G["root = manifest_dir/src\ncache in source_roots"]
    F -- No --> H["Tier 2: closure_donor_source_root(path)"]
    H --> I{any memoized entry\nwhose closure contains path?}
    I -- Yes --> J["root = donor.source_root()\ncache in source_roots"]
    I -- No --> K["Tier 3: path.parent()\n(own directory, NOT cached)"]

    G --> L["FileAnalysis::compute(vfs, path, root, gen)"]
    J --> L
    K --> L
    USE --> L
    L --> M["store in analyses map"]
    M --> R

    subgraph close_document
    N["source_roots.remove(path)"] --> O["invalidate(path, false)"]
    end
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A["analysis(path) called"] --> B{analysis memoized?}
    B -- Yes --> R["return cached FileAnalysis"]
    B -- No --> C["resolve_source_root(path)"]

    C --> D{sticky root cached\nin source_roots?}
    D -- Yes --> USE["use cached root"]

    D -- No --> E["Tier 1: manifest_source_root(path)"]
    E --> F{Inference.toml found\n& file under src/?}
    F -- Yes --> G["root = manifest_dir/src\ncache in source_roots"]
    F -- No --> H["Tier 2: closure_donor_source_root(path)"]
    H --> I{any memoized entry\nwhose closure contains path?}
    I -- Yes --> J["root = donor.source_root()\ncache in source_roots"]
    I -- No --> K["Tier 3: path.parent()\n(own directory, NOT cached)"]

    G --> L["FileAnalysis::compute(vfs, path, root, gen)"]
    J --> L
    K --> L
    USE --> L
    L --> M["store in analyses map"]
    M --> R

    subgraph close_document
    N["source_roots.remove(path)"] --> O["invalidate(path, false)"]
    end
```

</a>

<sub>Reviews (3): Last reviewed commit: ["Merge branch &#39;main&#39; into 243-bug-fix-sta..."](https://github.com/inferara/inference/commit/fc83f883c22ecb420b28215f71bce8c73e31a6b3) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=44787991)</sub>

### CODE REVIEW

File: ide/ide-db/src/database.rs [hunk 1]
```diff
@@ -94,15 +117,59 @@ impl RootDatabase {
...
+    /// wins, so the choice is deterministic across repeated analyses.
+    fn closure_donor_source_root(&self, file: &Path) -> Option<PathBuf> {
+        self.analyses
+            .iter()
+            .filter(|(entry, analysis)| {
+                entry.as_path() != file && analysis.closure_contains(file)
+            })
+            .min_by(|(a, _), (b, _)| a.cmp(b))
+            .map(|(_, analysis)| analysis.source_root().to_path_buf())
+    }
```
[the remark this answers is not in the document]
Comment (0xGeorgii) [reply ide/ide-db/src/database.rs:208]: Fixed. The resolved source root is now sticky per open document: RootDatabase caches a definitive root (manifest or closure donor) keyed by the entry path and reuses it on every recompute, so an adopted donor root survives the donor's eviction. Re-analyzing the non-entry file first after a shared-dependency change now reuses the borrowed root instead of falling to its own directory. The own-directory fallback is deliberately left uncached so a later-opened governing entry can still upgrade the root, and didClose drops the cached root so the next open re-resolves from scratch. Added the missing coverage: closure_donor_root_survives_shared_dep_change_when_dependent_reanalyzes_first (donor and dependent both evicted by a shared-dep change, dependent re-analyzed first, no false import diagnostics) and closing_a_document_drops_its_sticky_source_root.
